pkg/storage/ibmcos/ibmcos.go

```diff
@@ -0,0 +1,750 @@
+	}
+
+	// Get resource manager service
+	rm, err := d.getResourceManagerService()
+	if err != nil {
+		return err
+	}
+
+	// Check if service instance exists
+	if len(d.Config.ServiceInstanceCRN) != 0 {
+		instance, resp, err := rc.GetResourceInstanceWithContext(
+			d.Context,
+			&resourcecontrollerv2.GetResourceInstanceOptions{
+				ID: &d.Config.ServiceInstanceCRN,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("unable to get resource instance: %s with resp code: %d", err.Error(), resp.StatusCode)
+		}
+
+		switch *instance.State {
+		case resourcecontrollerv2.ListResourceInstancesOptionsStateActiveConst:
+			// Service instance exists and is active
+			if *instance.ResourceGroupID != "" {
+				// Get resource group name
+				rg, resp, err := rm.GetResourceGroupWithContext(
+					d.Context,
+					&resourcemanagerv2.GetResourceGroupOptions{
+						ID: instance.ResourceGroupID,
+					},
+				)
+				if err != nil {
+					return fmt.Errorf("unable to get resource group: %s with resp code: %d", err.Error(), resp.StatusCode)
+				}
+				// Set resource group name
+				d.Config.ResourceGroupName = *rg.Name
+			}
+			cr.Status.Storage = imageregistryv1.ImageRegistryConfigStorage{
+				IBMCOS: d.Config.DeepCopy(),
+			}
+			cr.Spec.Storage.IBMCOS = d.Config.DeepCopy()
+			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionFalse, "IBM COS Instance Active", "IBM COS service instance is active")
+		case resourcecontrollerv2.ListResourceInstancesOptionsStateProvisioningConst:
+			// Service instance exists and is provisioning
+			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionFalse, "IBM COS Instance Provisioning", "IBM COS service instance is provisioning")
+			return fmt.Errorf("waiting for IBM COS service instance to finish provisioning")
+		default:
+			// Service instance does not exist, will create one
+			d.Config.ServiceInstanceCRN = ""
+			util.UpdateCondition(cr, defaults.StorageExists, operatorapi.ConditionFalse, "IBM COS Instance Gone", "IBM COS service instance is inactive or has been removed.")
+		}
+	}
+
+	// Attempt to create a new service instance
+	if len(d.Config.ServiceInstanceCRN) == 0 {
+		// Get resource group details
+		resourceGroups, resp, err := rm.ListResourceGroupsWithContext(
+			d.Context,
+			&resourcemanagerv2.ListResourceGroupsOptions{
+				Name: &d.Config.ResourceGroupName,
+			},
+		)
+		if resourceGroups == nil || err != nil {
+			return fmt.Errorf("unable to get resource groups: %s with resp code: %d", err.Error(), resp.StatusCode)
+		} else if len(resourceGroups.Resources) == 0 {
+			return fmt.Errorf("unable to find any resource groups with resp code: %d", resp.StatusCode)
+		}
+
+		// Define instance options
+		serviceInstanceName := fmt.Sprintf("%s-%s", infra.Status.InfrastructureName, defaults.ImageRegistryName)
+		serviceTarget := "bluemix-global"
+		resourceGroupID := *resourceGroups.Resources[0].ID
+		resourcePlanID := "744bfc56-d12c-4866-88d5-dac9139e0e5d"
```

Comment:
Is it some well-known ID?

Reply:
Yes, it is the COS service plan ID on IBM Cloud catalog.